features/256-auth-gear-interaction-overview-in-next/api.md

```diff
@@ -0,0 +1,402 @@
+2. `context.req.body`: original request body from client, ex: for login, it would be an object with username and password.
+3. `context.req.id`: original request ID.
+4. `context.secrets`: secrets of the hook.
+
+## user common properties and user profile
+
+For future advanced management requirements, auth gear should have user common properties, which is saved for common user properties, such as avatar, first name, last name, display name, preferred language, ..., etc. 
+
+User common properties would be great help for better auth gear use experience, which allows to provide API response in preferred language, segment support, multi-lang custom email template.
+
+User profile is for used for varied user properties, they differ from application to application, such as: ethnic, height, weight, hobby,..., etc.
+
+```
+CREATE TABLE _auth_user_profile (
+  user_id text REFERENCES _core_user(id),
+  created_at timestamp without time zone NOT NULL,
+  created_by text,
+  updated_at timestamp without time zone NOT NULL,
+  updated_by text,
+  data jsonb,
+  PRIMARY KEY(user_id),
+  UNIQUE (user_id)
+);
+
+CREATE TABLE _core_user (
+  id text PRIMARY KEY,
+  token_valid_since timestamp without time zone,
+  
+  last_seen_at timestamp without time zone,
+  last_login_at timestamp without time zone,
+  
+  disabled boolean NOT NULL DEFAULT false,
+  disabled_message text,
+  disabled_expiry timestamp without time zone,
+  
+  verified boolean NOT NULL DEFAULT false,
+  verify_info jsonb NOT NULL DEFAULT '{}'::JSONB,
+  
+  avatar_url text,
+  first_name text,
+  last_name text,
+  display_name text,
+  birthday imestamp without time zone,
+  gender text,
+  prefer_lang_id text REFERENCES _core_lang(id),
+  ...
+);
+```
+
+The structure of user attributes could be:
+
+```
+{
+    id: <id>,
+    createdAt: <createdAt>,
+    updatedAt: <updatedAt>,
+    disabled: <disabled>,
+    roles: [<role>, <role>, <role>, ...],
+    avatarUrl: <avatarUrl>,
+    birthday: <birthday>,
+    preferredLang: <preferredLang>,
+    // more common user attributes
+    ...
+    profile: {
+        // any other free form data
+        ...
+    }
+}
+```
+
+## API gateway should add “current user” in request context
+
+Since Cloud Function would indicate it is executed by authenticated user only, so API gateway may handle the authentication process, and generate "current user" context for both auth gear and Cloud Function. (Currently, API gateway didn't handle user authentication process, it is handled by auth gear middleware)
+
+![](https://i.imgur.com/ohlsXTW.png)
+1. SDK send Cloud Function request to API gateway.
+2. API gateway check if the user is authenticated.
+3. If the user is authenticated, API gateway routes the request to Cloud Function.
+
+```javascript=
+const skygear = require('skygear');
+
+module.exports = (req, res) => {
+    res.end("welcome " + req.context.user.email);
+}
+```
+
+## Use cases: save user profile in Cloud Function DB after user signup
```

Comment:
I think we should write a short discussion about why this use case is important here. Maybe something like these?

"We decided the recommended use cases, is user of Skygear Auth shall use `after_user_profile_and_metadata_updated` to create and sync a user profile table in their own database if they need it, below is an example of how it should be done"

Reply:
BTW, the hook name is fake... 

Moreover, maybe we should provide / merge the sample of "create" and update? and also show that we will have an API to let user update the metadata and profile in ONE example?